libatalk/adouble/ad_attr.c

```diff
@@ -86,62 +100,122 @@ int ad_setattr(const struct adouble *ad, const uint16_t attribute)
 /* --------------
  * save file/folder ID in AppleDoubleV2 netatalk private parameters
  * return 1 if resource fork has been modified
+ * return -1 on error.
  */
 int ad_setid (struct adouble *adp, const dev_t dev, const ino_t ino , const uint32_t id, const cnid_t did, const void *stamp)
 {
     uint32_t tmp;
+    char *ade = NULL;
 
     ad_setentrylen( adp, ADEID_PRIVID, sizeof(id));
     tmp = id;
     if (adp->ad_vers == AD_VERSION_EA)
         tmp = htonl(tmp);
-    memcpy(ad_entry( adp, ADEID_PRIVID ), &tmp, sizeof(tmp));
+
+    ade = ad_entry(adp, ADEID_PRIVID);
+    if (ade == NULL) {
+        LOG(log_warning, logtype_ad, "ad_setid: failed to set ADEID_PRIVID\n");
+        return -1;
+    }
```

Comment:
I found an issue introduced here, where if you copy a file into the shared directory on the Netatalk host while a Mac client is connected, afpd will throw an error here and exit. There may be a deeper issue, but I suggest we remove this validation for now.

```
Apr 18 10:16:37 rasp32 afpd[13533]: ad_setid: failed to set ADEID_PRIVID
Apr 18 10:16:38 rasp32 afpd[13533]: ad_getid: failed to retrieve ADEID_PRIVDEV
```

Reply:
If we skip this validation here we'll NULL deref and crash in next line. It's better to figure out why `ad_entry()` is returning NULL here.

Reply:
One possibility is that in ad_open() we lstat() file and set valid_data_len to `0` if it fails. Maybe add a log statement there and see if that's why we're getting an invalid adouble.

Reply:
Right, so I was thinking of making the successful check a criteria for doing the memcpy. Something like:

```
--- a/libatalk/adouble/ad_attr.c
+++ b/libatalk/adouble/ad_attr.c
@@ -148,9 +148,14 @@ int ad_setid(struct adouble *adp, const dev_t dev, const ino_t ino,
         ade = ad_entry(adp, ADEID_PRIVID);
         if (ade == NULL) {
             LOG(log_warning, logtype_default, "ad_setid: failed to set ADEID_PRIVID\n");
-             return -1;
-        }
-        memcpy(ade, &id, sizeof(id));
+        } else
+            memcpy(ade, &id, sizeof(id));

                ad_setentrylen(adp, ADEID_PRIVDEV, sizeof(dev_t));
                if ((adp->ad_options & ADVOL_NODEV)) {
```

I haven't tried adding logging yet, but from a black box perspective it's pretty obvious what happens though, isn't it? A newly created file on the host file system simply doesn't have any adouble metadata yet. It does get created some time after ad_setid() has been called so the next you connect to the same afpshare, the error won't happen again. Perhaps adding a check from wherever ad_setid() would make sense?

Reply:
Existing debug logging gives this (after doing 'touch 123' and then opening the volume with the 123 file on the Mac client side.)

```
Apr 18 16:22:56.522879 afpd[14082] {file.c:631} (D5:AFPDaemon): getfilparams("123")
Apr 18 16:22:56.522999 afpd[14082] {ad_open.c:1500} (D5:Default): ad_open(".AppleDouble/123"): {cwd: "/home/dmark/afpshare"} creating adouble file
Apr 18 16:22:56.523275 afpd[14082] {file.c:317} (D5:AFPDaemon): getmetadata("123")
Apr 18 16:22:56.523340 afpd[14082] {dircache.c:384} (D5:AFPDaemon): dircache_search_by_name(did:2, "123")
Apr 18 16:22:56.523403 afpd[14082] {dircache.c:417} (D5:AFPDaemon): dircache(did:2,"123"): {not in cache}
Apr 18 16:22:56.523494 afpd[14082] {ad_attr.c:217} (W:Default): ad_getid: failed to retrieve ADEID_PRIVDEV

Apr 18 16:22:56.523562 afpd[14082] {cnid_dbd.c:577} (D5:CNID): cnid_dbd_add: CNID: 2, name: '123', inode: 0x218b8, type: 0 (0=file, 1=dir)
Apr 18 16:22:56.523758 afpd[14082] {cnid_dbd.c:228} (D10:CNID): write_vec: wrote 43 bytes
Apr 18 16:22:56.523823 afpd[14082] {cnid_dbd.c:290} (D10:CNID): send_packet: {done}
Apr 18 16:22:56.526082 afpd[14082] {cnid_dbd.c:355} (D10:CNID): dbd_rpc: {done}
Apr 18 16:22:56.526205 afpd[14082] {cnid_dbd.c:409} (D10:CNID): transmit: {done}
Apr 18 16:22:56.526270 afpd[14082] {cnid_dbd.c:589} (D5:CNID): cnid_dbd_add: got CNID: 17302
Apr 18 16:22:56.526344 afpd[14082] {file.c:286} (D5:AFPDaemon): get_id(/home/dmark/afpshare/123): calling ad_setid(old: 0, new: 17302)
Apr 18 16:22:56.526406 afpd[14082] {ad_attr.c:148} (W:Default): ad_setid: failed to set ADEID_PRIVID

Apr 18 16:22:56.526483 afpd[14082] {file.c:336} (D5:AFPDaemon): getmetadata: caching: did:2, "123", cnid:17302
Apr 18 16:22:56.526585 afpd[14082] {dircache.c:494} (D5:AFPDaemon): dircache(did:17302,'123'): {added}
Apr 18 16:22:56.526692 afpd[14082] {fault.c:122} (S:Default): ===============================================================
Apr 18 16:22:56.526757 afpd[14082] {fault.c:123} (S:Default): INTERNAL ERROR: Signal 11 in pid 14082 (2.2.6)
Apr 18 16:22:56.526817 afpd[14082] {fault.c:124} (S:Default): ===============================================================
Apr 18 16:22:56.527984 afpd[14082] {fault.c:96} (S:Default): BACKTRACE: 0 stack frames:
```

So never mind my assumption; it's clearly creating the metadata first.

Reply:
@anodos325 As mentioned in my email, this issue turns out not to affect Netatalk 3.1 after all. Please disregard this as far as holding back the merging into master/3.1. I will keep working on this over at my Netatalk 2.2 backport PR instead.